### PR TITLE
[SPARK-31129][SQL][Tests] Fix IntervalBenchmark and DateTimeBenchmark

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR aims to recover `IntervalBenchmark` and `DataTimeBenchmark` due to banning intervals as output.

### Why are the changes needed?

This PR recovers the benchmark suite.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Manually, re-run the benchmark.

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan, thank you very much.

**Comment by dongjoon-hyun** [comment]
Hi, @yaooqinn . If this is not the last one, could you proceed as a single PR for all benchmark?

**Comment by dongjoon-hyun** [comment]
Also, please create a new JIRA issue. The old one is already committed as `ExtractBenchmark`. So, we cannot reuse the ID for `IntervalBenchmark`.

**Comment by yaooqinn** [comment]
> Hi, @yaooqinn . If this is not the last one, could you proceed as a single PR for all benchmark?

I'll re-run all benchmark to see if any same issue exists, thank you very much  for your advice

**Comment by yaooqinn** [comment]
### org.apache.spark.sql.execution.benchmark.DateTimeBenchmark
```scala
[error] Caused by: java.time.format.DateTimeParseException: Text '2019-01-27 11:02:01.0' could not be parsed at index 20
[error] 	at java.base/java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:2046)
[error] 	at java.base/java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1874)
[error] 	at org.apache.spark.sql.catalyst.util.Iso8601TimestampFormatter.$anonfun$parse$1(TimestampFormatter.scala:71)
[error] 	... 19 more
```

**Comment by cloud-fan** [comment]
@yaooqinn can you point out what query causes this issue?

**Comment by yaooqinn** [comment]
1. the IntervalBenchmark is as same as the last one in ExtractBenchmark
2. the DateTimeBenchmark fails at `to_timestamp('2019-01-27 11:02:01.0', 'yyyy-MM-dd HH:mm:ss.SSS')`, which maybe should be `to_timestamp('2019-01-27 11:02:01.000', 'yyyy-MM-dd HH:mm:ss.SSS')` @cloud-fan

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala [hunk 1]
```diff
@@ -114,12 +114,12 @@ object DateTimeBenchmark extends SqlBasedBenchmark {
     }
     runBenchmark("Parsing") {
       val n = 1000000
-      val timestampStrExpr = "concat('2019-01-27 11:02:01.', cast(mod(id, 1000) as string))"
+      val timestampStrExpr = "concat('2019-01-27 11:02:01.', rpad(mod(id, 1000), 3, '0'))"
       val pattern = "'yyyy-MM-dd HH:mm:ss.SSS'"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala]: one more thought: maybe we should change the pattern string to `yyyy-M-d H:m:s.S`, to match the behavior of 2.4

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/DateTimeBenchmark.scala]: It will fail with `2019-01-27 11:02:01.12(123)`

**Comment by cloud-fan** [comment]
A high-level comment: "Optimize Local Shuffle Reader" should be a sub-feature of converting SMJ to BHJ. We should mention that instead.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for recovering benchmark suites, @yaooqinn and @cloud-fan .
I also checked the failure and the recovery. Merged to master/3.0.

**Pull request closed without merging**